keps/sig-release/20190316-rebase-images-to-distroless.md

```diff
@@ -0,0 +1,149 @@
+## Motivation
+
+Rebasing the k8s images to distroless can make the images thinner, safer and less vulnerable.
+
+Meanwhile, it will drastically reduce churn on the total number of k8s images versions. Due to the fact that many images are based on debian base and a vulnerability in debian base (a couple times a month) will result in rebuilding every image, changing the image from debian base to distroless static can reduce the total number of k8s images.
+
+What's more, it reduces the burden of managing and maintaining multiple k8s images from the security (e.g. CVE), compatibility and build process concerns.
+### Goals
+
+Use image gcr.io/distroless/static:latest as the only base image for the following K8s/GKE images
+
+- Images based FROM scratch
+- Images based on debian/alpine and only for the purpose of redirecting logs with shell.
+- Images based on debian due to previous rebasing from busybox.
+
+Document the full list of distroless-based images as well as the exception list
+### Non-Goals
+
+- Do not change Images based on debian/alpine that requires fluentd (e.g. hyperkube).
+- Do not change images that are based on debian-iptable images.
+## Background
+
+This section discusses how the goal and scope are determined due to the reality. It also contains the real use cases.
+### Kubernetes Images
+
+Kubernetes not only runs images in the containers, but its components themselves are running and deploying as images. Each component image can be built from different base images.
+
+Currently, kubernetes uses three main types of base images to build their components.
+
+#### Type 1 FROM Scratch
+This docker image is based “FROM scratch” and doesn’t have external dependencies. The original motivation of using “FROM scratch” is to keep the image extremely thin and only contain what a static binaries need. However, caveats are found when running the go static binaries due to some missing non-binary dependencies like ca-certificates, resolv.conf, hosts, and nsswitch (see [issue/69195](https://github.com/kubernetes/kubernetes/issues/69195)).
+#### Type 2 Debian Based Images
+An image can be based from Debian due to different reasons.
+
+- One big reason is that the image needs to use shell to redirect the glog. This base image now is an overkill because K8s 1.13 can support using klog which accepts a --log-file flag to point to the log path directly. (Historically images doing this mostly relied on busybox or alpine. Some recent change has migrated off those to debian-base. [PR/70245](https://github.com/kubernetes/kubernetes/pull/70245/files))
+- Another reason of using debian is from the CVE concerns. Those images are originally rebased from busybox to debian for better CVE feeds and management (See [PR/70245](https://github.com/kubernetes/kubernetes/pull/70245)).
+- A third type of images uses debian for certain external dependencies. For example, kube-proxy has a hard requirement for dynamic-link which uses glibc.
```

Comment:
nit: this isn't quite accurate. kube-proxy has a hard requirement on iptables. One option would be to build iptables from source, but then the burden is on us to keep it patched. We decided it's less management to let debian do the patching, and then we just need to rev the image when it needs to be patched.

Reply:
right, `kube-proxy` is not dynamically linked. (I think the only core binaries that need dynamic linking are `kubelet` and `hyperkube`; `kubelet` doesn't run in a docker image, and `hyperkube` is out of scope for this KEP.)